ehsm_kms_service/function.js

```diff
@@ -0,0 +1,273 @@
+
+/**
+ * Verify each parameter in the payload, such as data type,
+ * data length, whether it is the specified value,
+ * whether it is required, etc.
+ * For parameter description, see <ehsm_kms_params.js>
+ * @param {object} req
+ * @param {object} res
+ * @returns true | false
+ */
+const _checkPayload = function (req, res) {
+  try {
+    const action = req.query.Action
+    const { payload } = req.body
+    const currentPayLoad = ehsm_kms_params[action]
+    for (const key in currentPayLoad) {
+      if (
+        (payload[key] == undefined || payload[key] == '' || !payload[key]) &&
+        currentPayLoad[key].required
+      ) {
+        res.send(_result(400, 'Missing required parameters'))
+        return false
+      }
+      if (currentPayLoad[key].type == 'string' && payload[key]) {
+        if (typeof payload[key] != 'string') {
+          res.send(_result(400, `${key} must be of string type`))
+          return false
+        }
+        if (
+          payload[key] != undefined &&
+          ((currentPayLoad[key].maxLength &&
+            payload[key].length > currentPayLoad[key].maxLength) ||
+            (currentPayLoad[key].minLength &&
+              payload[key].length < currentPayLoad[key].minLength))
+        ) {
+          res.send(_result(400, `${key} length error`))
+          return false
+        }
+      }
+      if (currentPayLoad[key].type == 'int' && payload[key]) {
+        if (!Number.isInteger(payload[key])) {
+          res.send(_result(400, `${key} must be of integer type`))
+          return false
+        }
+        if (
+          payload[key] != undefined &&
+          ((currentPayLoad[key].maxNum &&
+            payload[key] > currentPayLoad[key].maxNum) ||
+            (currentPayLoad[key].minNum &&
+              payload[key] < currentPayLoad[key].minNum))
+        ) {
+          res.send(
+            _result(
+              400,
+              `${key} must be between ${currentPayLoad[key].minNum} and ${currentPayLoad[key].maxNum}`
+            )
+          )
+          return false
+        }
+      }
+      if (currentPayLoad[key].type == 'const' && payload[key]) {
+        if (!currentPayLoad[key].arr.includes(payload[key])) {
+          res.send(
+            _result(400, currentPayLoad[key].errortext || `${key} error`)
+          )
+          return false
+        }
+      }
+    }
+    return true
+  } catch (error) {
+    res.send(_result(400, 'Parameter exception', {}))
+    logger.error(JSON.stringify(error))
+    return false
+  }
+}
+
+/**
+ * check params
+ */
+const _checkParams = function (req, res, next, nonce_database, DB) {
+  try {
+    const ACTION = req.query.Action
+
+    let ip = req.ip
+    if (ip.substr(0, 7) == '::ffff:') {
+      ip = ip.substr(7)
+    }
+    const _logData = {
+      body: req.body,
+      query: req.query,
+      ip,
+    }
+    logger.info(JSON.stringify(_logData))
+    if (ACTION === apis.RA_GET_API_KEY || ACTION === apis.RA_HANDSHAKE_MSG0) {
+      next()

```

Comment:
is it a todo ?

Reply:
RA_GET_API_KEY 、RA_HANDSHAKE_MSG0  these two interfaces do not have appid,, timestamp, sign, payload parameter checks